distributed/shuffle/_buffer.py

```diff
@@ -0,0 +1,245 @@
+from __future__ import annotations
+
+import asyncio
+import contextlib
+import logging
+from collections import defaultdict
+from collections.abc import Iterator
+from typing import TYPE_CHECKING, Any, Generic, Sized, TypeVar
+
+from distributed.metrics import time
+from distributed.shuffle._limiter import ResourceLimiter
+
+if TYPE_CHECKING:
+    import pyarrow as pa
+
+logger = logging.getLogger("distributed.shuffle")
+
+ShardType = TypeVar("ShardType", bound=Sized)
+
+
+class ShardsBuffer(Generic[ShardType]):
+    """A buffer for P2P shuffle
+
+    The objects to buffer are typically bytes belonging to certain shards.
+    Typically the buffer is implemented on sending and receiving end.
+
+    The buffer allows for concurrent writing and buffers shards to reduce overhead of writing.
+
+    The shards are typically provided in a format like::
+
+        {
+            "bucket-0": [b"shard1", b"shard2"],
+            "bucket-1": [b"shard1", b"shard2"],
+        }
+
+    Buckets typically correspond to output partitions.
+
+    If exceptions occur during writing, the buffer is automatically closed. Subsequent attempts to write will raise the same exception.
+    Flushing will not raise an exception. To ensure that the buffer finished successfully, please call `ShardsBuffer.raise_on_exception`
+    """
+
+    shards: defaultdict[str, list[ShardType]]
+    sizes: defaultdict[str, int]
+    _exception: None | Exception
```

Comment:
nit: Complete annotations for attributes
```suggestion
    shards: defaultdict[str, list[ShardType]]
    sizes: defaultdict[str, int]
    concurrency_limit: int
    memory_limiter: ResourceLimiter | None
    diagnostics: dict[str, float]
    max_message_size: int

    bytes_total: int
    bytes_memory: int
    bytes_written: int
    bytes_read: int

    _closed: bool
    _done: bool
    _exception: None | Exception
    _tasks: list[asyncio.Task]
    _shards_available: asyncio.Condition
    _flush_lock: asyncio.Lock
```